### PR TITLE
logging: Add optional function name prefix

### LINKED ISSUE #10644: logging: Where's my function name??
With new logger, I get:

~~~
[00:00:19.606,640] <err> net_tcp: context->tcp == NULL
[00:00:19.612,784] <err> net_tcp: context->tcp == NULL
[00:00:19.615,296] <err> net_tcp: context->tcp == NULL
[00:00:19.618,140] <err> net_tcp: context->tcp == NULL
~~~

What's this? How to get use of this? I need the function name where it happens, like was there with the old logger. (It was there, right, or am I hallucinating? All in all, with the old logger, it was clear (ahem) what happens where, and the new logger really leads to that question - what's this?)

### LINKED ISSUE #10700: logging: Support for including function name in the log
I'd like to proceed with converting all BT_* logging calls to corresponding LOG_* ones as part of #10264 however e.g. BT_DBG and BT_ERR implicitly include the function name in the log output, so this would change the current behavior. Could we either get a Kconfig option to include the function name, or possibly store it as part of log_msg_hdr (as a string pointer) so that the log processing can later decide to include it or not in the output.

@carlescufi @nashif @jukkar FYI

### PR DESCRIPTION
**Pull request by nordic-krch**
Extended logger to support optional log message prepending with function name.

Prepending with function name is optional (enabled by default). I've used following formatting:
`[00:00:24.229,644] <err> app_test.foo(): err message`

Fixes #10644
Fixes #10700 

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

### CODE REVIEW

File: include/logging/log_core.h
Comment (jhedberg) [inline include/logging/log_core.h]: This added line seems unrelated to the PR (it's also redundant)

File: subsys/logging/log_output.c
Comment (jhedberg) [inline subsys/logging/log_output.c]: Just a minor whitespace fix to make this more readable: add a space between the : and the opening " of the second string.

**Review by jhedberg — APPROVED** [review]
Looks generally good, just a few coding style comments

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: include/logging/log_core.h
Comment (nordic-krch) [reply include/logging/log_core.h]: i was thinking that () will be self explained but if you think that it's not worth spending 2 characters i will remove it.

Comment (pizi-nordic) [reply include/logging/log_core.h]: I would keep keep the parentheses.

**Comment by nordic-krch** [comment]
it seems that because i added one additional parameter i need to extend logger to support 10 arguments and it is not straight forward since 10 arguments don't fit in 2 chunks of log message. Need to extend that.

File: include/logging/log_core.h
Comment (jhedberg) [reply include/logging/log_core.h]: @pizi-nordic I don't have a particularly strong opinion either way, just the general desire to keep the messages as compact as possible without sacrificing readability.

**Review by jhedberg — APPROVED** [review]

**Comment by jukkar** [comment]
> Once this PR from @jukkar is merged there should still only be 9 arguments required (unless I'm overlooking something non-networking related): #10711

I think #10711 should solve this argument count issue in networking side. Someone just needs to approve that change, after that I can merge it.

**Comment by jukkar** [comment]
#10711 is now merged, @nordic-krch you could now rebase against master and re-submit

**Review by pizi-nordic — APPROVED** [review]

File: include/logging/log_core.h
Comment (nordic-krch) [reply include/logging/log_core.h]: @jhedberg i've noticed that #10711 @jukkar added %s() in case CONFIG_LOG_FUNCTION_NAME. I'm also leaning towards keeping parentheses (self-explaining as the argument).

Comment (pfalcon) [reply include/logging/log_core.h]: >  I'm also leaning towards keeping parentheses (self-explaining as the argument).

IMHO, drop it and save bytes. Likewise, intend to be propose to drop xml-like syntax for `<err>` vs `ERR`. (On critical path comparing that to the old log output.)

Comment (nordic-krch) [reply include/logging/log_core.h]: Seems no approach to suit them all. However, i'm more convinced to change <err> to ERR that removing brackets. There is also instance level filtering. In that case messages are prefixed with <module_name>.<instance_name>.<function_name> (e.g. `foo.my_foo.init`) with those 3 words together brackets improve readability otherwise function name may be confused with instance name. Unless we use something else than `.` as the separator in one of those cases. but it has to be something that is not used in naming (`~`, `#` ?).

Some examples
- `foo.my_foo#init` and  `foo#init`
- `foo~my_foo.init` and `foo.init`

No strong opinion but imo any of those examples is less readable than `foo.my_foo.init()`.

Comment (jukkar) [reply include/logging/log_core.h]: > No strong opinion but imo any of those examples is less readable than `foo.my_foo.init()`

Using dots to separate the fields looks less crowded and easier to read than the other options.

**Review by dbkinder — APPROVED** [review]
+1 for doc change

**Comment by nordic-krch** [comment]
Rebased on #10794 which adds support to 10 arguments in log api.

**Review by finikorg — APPROVED** [review]

**Pull request merged**
